### DISCUSSION TITLE
In the input file, how can I take a moment using values from another moment?

### DISCUSSION THREAD

**Discussion by moisesa2** [BLAST-WarpX/warpx#4303] [Data Analysis]
I am trying to calculate the heat flux given which is defined as 

q = integral { m /2 (vx-Vx) |**v-V**|^2 }fdx 

where Vx and **V** require a moment calculation itself. 

When I try to use Vx which I have defined as flow, the second moment (q) does not run and I get the following error

```
amrex::Abort::26::
### ERROR   : makeParser::Unknown symbol flow
#            (/home/xpa/workspace/builds/const_source_v2/WarpX/Source/Utils/Parser/ParserUtils.cpp:144)
 !!!
SIGABRT
```

**Comment by RemiLehe** [comment]
Thanks a lot for raising this issue.

Would you be willing to post your full input script (or a modified version thereof that still reproduces the above error)? I think it would help understand where the problem is coming from.

**Reply by moisesa2** [reply]
The following is the section of the input script that handles diagnostic output

diagnostics.diags_names = diag diag_full diag_check

######################### Field data
diag.intervals = 500
diag.diag_type = Full
diag.file_min_digits = 7
diag.fields_to_plot = Ex Ey rho rho_electrons rho_ions  phi rho_ions2
diag.plot_raw_fields = 1
diag.write_species = 0


diag.particle_fields_to_plot = kinetic flow flow_y flow_z q_i pi_i
diag.particle_fields_species = ions electrons ions2

#########momentum moment
diag.particle_fields.kinetic.do_average = 1
diag.particle_fields.kinetic(x,y,z,ux,uy,uz) = mi*ux**2


########Flow speeds
diag.particle_fields.flow.do_average = 1
diag.particle_fields.flow(x,y,z,ux,uy,uz) = ux

diag.particle_fields.flow_y.do_average = 1
diag.particle_fields.flow_y(x,y,z,ux,uy,uz) = uy

diag.particle_fields.flow_z.do_average = 1
diag.particle_fields.flow_z(x,y,z,ux,uy,uz) = uz

#########Heat flux
diag.particle_fields.q_i.do_average = 1
diag.particle_fields.q_i(x,y,z,ux,uy,uz) = (ux - flow)* ((ux-flow)**2 + (uy-flow_y)**2 + (uz-flow_z)**2)*mi*0.5


########Stress tensor (xx component)
diag.particle_fields.pi_i.do_average = 1
diag.particle_fields.pi_i(x,y,z,ux,uy,uz) = mi * ((ux - flow)**2 - ((ux-flow)**2 + (uy-flow_y)**2 + (uz-flow_z)**2)/3)


########## Particle data
diag_full.intervals = 20000
diag_full.diag_type = Full
diag_full.file_min_digits = 7
diag_full.fields_to_plot = Ex Ey rho rho_electrons rho_ions  phi rho_ions2
diag_full.plot_raw_fields = 1



############### Checkpoint data
diag_check.intervals = 50000
diag_check.diag_type = Full
diag_check.file_min_digits = 7
#diag_check.plot_raw_fields = 1
diag_check.format = checkpoint

**Reply by ax3l** [reply]
Currently, expressions like these
```
diag.particle_fields.q_i(x,y,z,ux,uy,uz) = (ux - flow)* ((ux-flow)**2 + (uy-flow_y)**2 + (uz-flow_z)**2)mi0.5
```
that reference another field (`flow` in this case) are not supported.

I would output fields:
- `flow`, `flow_y`, `flow_z`
- `ux`, `uy`, `uz`
and calculate the quantity of interest in post-processing.

Alternatively, try to express `q_i` and `pi_i` without referencing other fields. It looks like you depend on a spatial average of u_i for the point-wise quantities, so I am not sure that works here.

Alternatively, if you prefer in situ calculations, you can use our PICMI Python inputs with pyAMReX bindings, allocate another AMReX MultiFab from Python for your quantities of interest on the fly and calculate the results into that extra field on the fly:
- https://github.com/AMReX-Codes/pyamrex/
- https://pyamrex.readthedocs.io (Examples linked within, but not yet for your exact use case)

If you like to use Python for your simulations, then we can guide you how to achieve this.

**Reply by Kellysherry** [reply]
hi, I encountered a similar problem.
I want to define the incident particle density by local electrical field calculated last step, but it seems that this methode will cost lots of time and internal storage since I'll have to output and input the electrical field in every step?
Is there another way to call the data calculated during the simulation?

**Comment by dpgrote** [comment]
Can your expression be multiplied out so it is a function of the basic terms?
As an example, the RMS can be written
rms(v) = sqrt(ave((v - vbar)\**2)) = sqrt(ave(v\**2) - ave(v)**2)
This works well as long as the ave(v) is not too much greater than the rms(v).

**Reply by moisesa2** [reply]
In my case, the average velocity may exceed the rms velocity by a lot so this would not work I think.

**Reply by dpgrote** [reply]
Approximately, what is the ratio of ave/rms? For the rms calculation, the expanded version works well for ratios up to around 1.e6 (for double precision).
I'm asking since implementing what you want would be nontrivial. Presumably, the `flow` quantity would be the spatially dependent data calculated on the grid, so a grid interpolation would need to be done for each particle. Also, the parser currently does not allow extra parameters beyond the predefined ones like `x, y, z`, `my_constants`, and built in constants like `c`.
